src/paillier.rs

```diff
@@ -61,15 +63,58 @@ impl PaillierDecryptionKey {
         Ok(self.0.decrypt(c).ok_or(PaillierError::DecryptionFailed)?)
     }
 
-    /// Generate a new [`PaillierDecryptionKey`] from two primes.
+    /// Generate a new [`PaillierDecryptionKey`] and its factors.
     ///
-    /// The inputs `p` and `q` are checked for primality but not for _safe_ primality.
-    /// TODO #56: Check prime size and output size.
-    pub(crate) fn from_primes(p: &BigNumber, q: &BigNumber) -> Result<Self> {
-        Ok(PaillierDecryptionKey(
-            libpaillier::DecryptionKey::with_primes(p, q)
+    /// The factors `p` and `q` are `PRIME_BITS`-long safe primes, and the resulting
+    /// modulus is `2 * PRIME_BITS` long.
+    pub(crate) fn new(
+        rng: &mut (impl CryptoRng + RngCore),
+    ) -> Result<(Self, BigNumber, BigNumber)> {
+        // Local utility function to generate prime pairs and check their resulting modulus length.
+        let generate_prime_pair = || {
+            // As generating safe primes is very computationally expensive (> one minute per prime in github CI),
+            // we read precomputed ones from a file (but only in tests!)
+            #[cfg(not(test))]
+            let (p, q) = (
+                prime_gen::get_random_safe_prime_512(rng),
+                prime_gen::get_random_safe_prime_512(rng),
+            );
+            #[cfg(test)]
+            let (p, q) = prime_gen::get_prime_pair_from_pool_insecure(rng);
+
+            if (&p * &q).bit_length() == 2 * PRIME_BITS {
+                Some((p, q))
+            } else {
+                None
+            }
+        };
+
+        // A Paillier decryption key is the product of the two primes, but sometimes two n/2-bit primes can produce an
+        // n-1 bit modulus. Allow some retries to handle that error.
+        let (p, q) = std::iter::repeat_with(generate_prime_pair)
+            .take(CRYPTOGRAPHIC_RETRY_MAX)
+            .find(|result| result.is_some())
+            // We hit the maximum number of retries without getting an acceptable pair.
+            .ok_or(InternalError::RetryFailed)?
+            // The `find` function failed to actually give us `Some`.
+            .ok_or(InternalError::InternalInvariantFailed)?;
```

Comment:
Rust question: I want this to evaluate lazily so it only generates prime pairs until we get one that works. The `take` won't execute all 500 elements, right?

Reply:
I had to test it out with an example. Just from reading the doc, I thought it would always execute `CRYPTOGRAPHIC_RETRY_MAX` number of times. But yes, it is lazy!